srml/scored-pool/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+};
+
+type BalanceOf<T, I> = <<T as Trait<I>>::Currency as Currency<<T as system::Trait>::AccountId>>::Balance;
+
+pub trait Trait<I=DefaultInstance>: system::Trait {
+	/// The currency used for deposits.
+	type Currency: Currency<Self::AccountId> + ReservableCurrency<Self::AccountId>;
+
+	/// The score attributed to a member or candidate.
+	type Score: SimpleArithmetic + Clone + Copy + Default + Encode + Decode + MaybeSerializeDebug;
+
+	/// The overarching event type.
+	type Event: From<Event<Self, I>> + Into<<Self as system::Trait>::Event>;
+
+	// The deposit which is reserved from candidates if they want to
+	// start a candidacy. The deposit gets returned when the candidacy is
+	// withdrawn or when the candidate is kicked.
+	type CandidateDeposit: Get<BalanceOf<Self, I>>;
+
+	/// Every `Period` blocks the `Members` are filled with the highest scoring
+	/// members in the `Pool`.
+	type Period: Get<Self::BlockNumber>;
+
+	/// The receiver of the signal for when the membership has been initialized.
+	/// This happens pre-genesis and will usually be the same as `MembershipChanged`.
+	/// If you need to do something different on initialization, then you can change
+	/// this accordingly.
+	type MembershipInitialized: InitializeMembers<Self::AccountId>;
+
+	/// The receiver of the signal for when the members have changed.
+	type MembershipChanged: ChangeMembers<Self::AccountId>;
+
+	/// Allows a configurable origin type to set a score to a candidate in the pool.
+	type ScoreOrigin: EnsureOrigin<Self::Origin>;
+
+	/// Required origin for removing a member (though can always be Root).
+	/// Configurable origin which enables removing an entity. If the entity
+	/// is part of the `Members` it is immediately replaced by the next
+	/// highest scoring candidate, if available.
+	type KickOrigin: EnsureOrigin<Self::Origin>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait<I>, I: Instance=DefaultInstance> as ScoredPool {
+		/// The current pool of candidates, stored as an ordered Vec
+		/// (ordered descending by score, `None` last, highest first).
+		Pool get(pool) config(): Vec<(T::AccountId, Option<T::Score>)>;
+
+		/// The current membership, stored as an ordered Vec.
+		Members get(members): Vec<T::AccountId>;
+
+		/// Size of the `Members` set.
+		MemberCount get(member_count) config(): u32;
+	}
+	add_extra_genesis {
+		config(members): Vec<T::AccountId>;
+		config(phantom): sr_std::marker::PhantomData<I>;
+		build(|
+			storage: &mut (sr_primitives::StorageOverlay, sr_primitives::ChildrenStorageOverlay),
+			config: &Self,
+		| {
+			sr_io::with_storage(storage, || {
+				let mut pool = config.pool.clone();
+
+				// reserve balance for each candidate in the pool.
+				// panicking here is ok, since this just happens one time, pre-genesis.
+				pool
+					.iter()
+					.for_each(|(who, _)| {
+						T::Currency::reserve(&who, T::CandidateDeposit::get())
+							.expect("balance too low");
```

Comment:
Could we please get a better error message?

"Balance too low to submit candidacy" or something else.